docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+
+These changes would preserve the ability of operators to issue queries with
+curl (but would require using JSON-RPC instead of URI parameters). That would
+be a little less user-friendly, but for a use case that should not be that
+prevalent.
+
+These changes would also preserve compatibility with existing JSON-RPC based
+code paths like the `tendermint` CLI and the light client (even ahead of
+further work to remove that dependency).
+
+**Design goal:** An operator should be able to disable non-local access to the
+RPC server on any node in the network without impairing the ability of the
+network to function for service of state replication, including light clients.
+
+**Design principle:** All communication required to implement and monitor the
+consensus network should use P2P, including the various synchronizations.
+
+### Options for ABCI Transport
+
+The majority of current usage is in Go, and the majority of that is mediated by
+the Cosmos SDK, which uses the "direct call" interface. There is probably some
+opportunity to clean up the implementation of that code, notably by inverting
+which interface is at the "top" of the abstraction stack (currently it acts
+like an RPC interface, and escape-hatches into the direct call). However, this
+general approach works fine and doesn't need to be fundamentally changed.
+
+For applications _not_ written in Go, the two remaining options are the
+"socket" protocol (another variation on varint-prefiexed protobuf messages over
+an unstructured stream) and gRPC. It would be nice if we could get rid of one
+of these to reduce (unneeded?) optionality, and the "socket" protocol is the
+most obvious choice. gRPC is more complex, but is already widely used in the
+rest of the ecosystem (including the Cosmos SDK).
+
+The main question is whether requiring gRPC imposes an undue burden on the
+authors of applications in languages without good gRPC support. I propose that
+this is not a large constituency -- and if some important use case does arise
+later, it would not be too difficult for that application author to write a
+little proxy (in Go) that bridges the convenient SDK APIs into a simpler
+protocol.
+
+**Design principle:** It is better for an uncommon special case to carry the
+burdens of its specialness, than to bake an escape hatch into the infrastructure.
+
+**Recommendation:** We should deprecate and remove the socket protocol.
+
+### Options for RPC Transport
+
+[ADR 057][adr-57] proposes using gRPC for the Tendermint RPC implementation.
+This is still possible, but if we are able to simplify and decouple the
+concerns as described above, I do not think it should be necessary.
+
+While JSON-RPC is not the best possible RPC protocol for all situations, it has
+some advantages over gRPC for our domain. Specifically:
+
+- It is easy to call JSON-RPC manually from the command-line, which helps with
+  a common concern for the RPC service, local debugging and operations.
```

Comment:
Although it could be possible to build in query support in the `tendermint` CLI for gRPC to help node operators

Reply:
Yup, that's certainly a path we could take. It would tie operators to the Tendermint CLI a bit more (since writing your own gRPC tool is a lot more work than just curling the endpoint) but if we don't think the ad hoc usage is that important, this is definitely an option.

Reply:
It's worth noting that using JSON-RPC doesn't prevent us from shipping a CLI in the future. Using JSON-RPC does allow us to give a lot of flexibility and utility to operators without us having to maintain a CLI at the moment.  I do generally believe that there is a lot of upside in making an endpoint that is mostly interacted with by human operators easily human readable.